test/gtest/softmarginloss.hpp

```diff
@@ -0,0 +1,465 @@
+                                          "Error:"
+                                       << error << ",  Threshold: " << threshold;
+    }
+    SoftMarginLossTestCase softmarginloss_config;
+
+    tensor<T> input;
+    tensor<T> target;
+    tensor<T> dO;
+    tensor<T> dI;
+
+    tensor<T> ref_dI;
+
+    miopen::Allocator::ManageDataPtr input_dev;
+    miopen::Allocator::ManageDataPtr target_dev;
+    miopen::Allocator::ManageDataPtr dO_dev;
+    miopen::Allocator::ManageDataPtr dI_dev;
+};
+
+template <typename T = float>
+struct SoftMarginLossReducedForwardTest : public ::testing::TestWithParam<SoftMarginLossTestCase>
+{
+protected:
+    void SetUp() override
+    {
+        auto&& handle         = get_handle();
+        softmarginloss_config = GetParam();
+
+        auto in_dims    = softmarginloss_config.dims;
+        auto in_strides = softmarginloss_config.strides;
+
+        auto input_numel =
+            std::accumulate(in_dims.begin(), in_dims.end(), 1L, std::multiplies<int64_t>());
+        if(std::is_same<T, half_float::half>::value && input_numel > 80000)
+        {
+            std::cerr << "For fp16 forward reduction test, too many elements in input tensor can "
+                         "lead to fp16 "
+                         "overflow or underflow. If reduction mean, divisor is very big lead to "
+                         "underflow. If reduction sum, result is too big lead to overflow."
+                      << std::endl;
+            GTEST_SKIP();
+        }
+
+        auto gen_in_value = [](auto...) { return prng::gen_descreet_uniform_sign<T>(1e-2, 100); };
+        // below commented code that I have seen in many files will not work correctly with unpacked
+        // tensor tensor.generate() will call for_each() and this function only iterate through
+        // desc.GetLengths().size(), not desc.GetElementSpace() Example input_tensor to verify:
+        // input_tensor: dim(5, 3), stride(4, 1). Element space = 19, size = 15. Call above code
+        // will only generate value for 15 elements
+
+        // input = tensor<T>{in_dims, in_strides}.generate(gen_in_value);
+
+        // This is the right method to generate value for tensor
+        input = tensor<T>{in_dims, in_strides};
+        std::generate(input.begin(), input.end(), gen_in_value);
+
+        auto gen_target_value = [](auto...) {
+            return (prng::gen_A_to_B<int32_t>(0, 2) == 0) ? -1 : 1;
+        };
+        target = tensor<T>{in_dims, in_strides};
+        std::generate(target.begin(), target.end(), gen_target_value);
+
+        // Tensor with 1 element to store result after reduce
+        output = tensor<T>{std::vector<size_t>{1}};
+        std::fill(output.begin(), output.end(), 0);
+
+        ref_output = tensor<T>{std::vector<size_t>{1}};
+        std::fill(ref_output.begin(), ref_output.end(), 0);
+
+        ws_sizeInBytes = miopen::GetSoftMarginLossForwardWorkspaceSize(
+            handle, input.desc, target.desc, output.desc, 1);
```

Comment:
although not affect result, reduction param should be equal to divisor instead of 1